Allow datasets to be used in taskflow (#27540)
mistercrunch/Airflow: docs/apache-airflow/concepts/taskflow.rst

 .. Licensed to the Apache Software Foundation (ASF) under one

> or more contributor license agreements.  See the NOTICE file
> distributed with this work for additional information
> regarding copyright ownership.  The ASF licenses this file
> to you under the Apache License, Version 2.0 (the
> "License"); you may not use this file except in compliance
> with the License.  You may obtain a copy of the License at

Reviewer: ```suggestion
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at
```

 ..   http://www.apache.org/licenses/LICENSE-2.0

 .. Unless required by applicable law or agreed to in writing,

> software distributed under the License is distributed on an
> "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
> KIND, either express or implied.  See the License for the
> specific language governing permissions and limitations
> under the License.

Reviewer: ```suggestion
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.
```

Every logging line created this way will be recorded in the task log.

Passing Arbitrary Objects As Arguments
--------------------------------------

.. versionadded:: 2.5.0

As mentioned TaskFlow uses XCom to pass variables to each task. This requires that variables that are used as arguments
need to be able to be serialized. Airflow out of the box supports all built-in types (like int or str) and it
supports objects that are decorated with ``@dataclass`` or ``@attr.define``. The following example shows the use of
a ``Dataset``, which is ``@attr.define`` decorated, together with TaskFlow.

::

  Note: An additional benefit of using ``Dataset`` is that it automatically registers as an ``inlet`` in case it is used as an input argument. It also auto registers as an ``outlet`` if the return value of your task is a ``dataset`` or a ``list[Dataset]]``.

.. code-block:: python

    import json
    import pendulum
    import requests

    from airflow import Dataset
    from airflow.decorators import dag, task

    SRC = Dataset("https://www.ncei.noaa.gov/access/monitoring/climate-at-a-glance/global/time-series/globe/land_ocean/ytd/12/1880-2022.json")
    now = pendulum.now()

>     @dag(start_date=now, schedule="daily", catchup=False)

Reviewer 2: ```suggestion
    @dag(start_date=now, schedule="@daily", catchup=False)
```

Every logging line created this way will be recorded in the task log.

> Passing Arbitrary Objects As Parameters

Reviewer 2: ```suggestion
Passing Arbitrary Objects As Arguments
```

---------------------------------------

.. versionadded:: 2.5.0

> As mentioned TaskFlow uses XCom to pass variables to each task. This requires that variables that are used as parameters
> to tasks need to be able to be serialized. Airflow out of the box supports all built-in types (like int or str) and it
> supports objects that are decorated with ``@dataclass`` or ``@attr.define``, but it will need some help if you want to
> pass custom objects or if you want to control serialization yourself. To do so add the ``serialize()`` method to your
> class and the staticmethod ``deserialize(data: dict, version: int)`` to your class. Like so:

Reviewer 2: ```suggestion
TaskFlow uses XCom to pass variables to each task. This requires variables that are used as arguments
to tasks to be able to be serialized. Airflow out of the box supports all built-in types (like int or str) and it
supports objects that are decorated with ``@dataclass`` or ``@attr.define``, but it will need some help if you want to pass custom objects, or if you want to control serialization yourself. To do so, implement the ``serialize()`` method and ``deserialize(data: dict, version: int)`` staticmethod in your class. Like so:
```

Author: I *think* you are reformatting, but don't know exactly. Please reapply if you don't mind.

Reviewer 2: Did some rewording for clarity

Author: If you can do that again that would be appreciated, I am getting word blind here :-P

.. code-block:: python

    from typing import ClassVar

    class MyCustom:
        version: ClassVar[int] = 1

        def __init__(self, x):
            self.x = x

        def serialize(self) -> dict:
            return dict({"x": self.x})

        @staticmethod
        def deserialize(data: dict, version: int):
            if version > 1:
                raise TypeError(f"version > {MyCustom.version}")

>             return MyCustom(data["x"])

Reviewer 2: Could you demonstrate a taskflow task using the `MyCustom` class in the example?

Author: Are you open to "Tabbed" examples yet? :-)

Reviewer 2: Referring to https://github.com/apache/airflow/pull/26486 :-) ?

Author: Yes, but I am not ready for that yet :-). Example added.

Reviewer 2: I meant an example using this custom class. I'm not reading anything in relation to datasets. As a user, what do I do with `MyCustom`?
